src/shared/cloudformation/cloudformation.ts

```diff
@@ -284,4 +304,209 @@ export namespace CloudFormation {
+     */
+    export function getNumberForProperty(
+        property: string | number | object | undefined,
+        template: Template
+    ): number | undefined {
+        if (validatePropertyType(property, 'number', template)) {
+            if (typeof property === 'number') {
+                return property
+            } else if (isRef(property)) {
+                try {
+                    const forcedProperty = property as Ref
+                    return getReffedNumber(forcedProperty, template)
+                } catch (err) {
+                    getLogger().debug(err)
+                }
+            }
+        }
+
+        return undefined
+    }
+
+    /**
+     * Returns whether or not a property or its underlying ref matches the specified type
+     * Does not validate a default value for a template parameter; just checks the value's type
+     * @param property Property to validate the type of
+     * @param type Type to validate the property's type against
+     * @param template Template object to parse through
+     */
+    function validatePropertyType(
+        property: string | number | object | undefined,
+        type: 'string' | 'number',
+        template: Template
+    ): boolean {
+        if (typeof property === type) {
+            return true
+        } else if (isRef(property)) {
+            // property has a Ref, force it to abide by that shape
+            const forcedProperty = property as Ref
+            const param = getReffedParam(forcedProperty, template)
+            const paramType = param.Type === 'Number' ? 'number' : 'string'
+
+            return paramType === type
+        }
+
+        return false
+    }
+
+    /**
+     * Helper function to get a number from a Ref.
+     * Throws an error if the value is not specifically a number.
+     * Returns undefined if ref does not have a default value but is a number.
+     * @param ref Ref to pull a number from
+     * @param template Template to parse through
+     */
+    function getReffedNumber(ref: Ref, template: Template): number | undefined {
+        const param = getReffedParam(ref, template)
+        if (param.Type === 'Number') {
+            if (param.Default && typeof param.Default !== 'number') {
+                throw new Error(`Parameter ${ref.Ref} is not a number`)
+            }
+
+            // returns undefined if no default value is present
+            return param.Default
+        }
+
+        throw new Error(`Parameter ${ref.Ref} is not typed as a number`)
+    }
+
+    /**
+     * Helper function to get a string from a Ref.
+     * Throws an error if the value is specifically a number.
+     * (all other CFN param types are some form of string).
+     * Returns undefined if ref does not have a default value but is a string.
+     * @param ref Ref to pull a string from
+     * @param template Template to parse through
+     */
+    function getReffedString(ref: Ref, template: Template): string | undefined {
```

Comment:
nearly identical to `getReffedNumber`, they could share a generic implementation.

Reply:
See above. The problem would come in somewhere around line 434 since you'd have to determine the generic's type which is not doable at runtime in TS.

Reply:
I don't understand. Why wouldn't this work?

```typescript
function getReffedString(ref: Ref, template: Template): string | undefined {
    return getReffedThing(ref, template, 'string')
}
function getReffedNumber(ref: Ref, template: Template): number | undefined {
    return getReffedThing(ref, template, 'number')
}
function getReffedThing(
    ref: Ref,
    template: Template,
    thingType: 'number' | 'string'
    ): number | string | undefined
{
    const param = getReffedParam(ref, template)
    // every other type, including List<Number>, is formatted as a string.
    if ((thingType == 'number' && param.Type === 'Number')
        || (thingType != 'number' && param.Type !== 'Number')) {
        if (param.Default && typeof param.Default !== thingType) {
            throw new Error(`Parameter ${ref.Ref} is not a ${thingType}`)
        }

        // returns undefined if no default value is present
        return param.Default
    }

    throw new Error(`Parameter ${ref.Ref} is not a ${thingType}`)
}

```

Reply:
It does, I was fixating on a generic type mainly so we don't have to do a ` as string | undefined`. If you're cool with this this then I am too.

Reply:
It's a private function covered by tests so I think the DRY gain is worth the small type safety risk.